### PR TITLE
Make resource names( eg: storageclass, replicationcontroller etc) resolved before shortnames( eg: sc, rc etc) are resolved.

### PR DESCRIPTION
**Pull request by abdasgupta**
In reference with the following PR https://github.com/kubernetes/kubernetes/pull/43338 , this PR is raised. This PR makes the order of REST mapping to resource names -> shortnames  instead of current shortnames -> resource names

cc @deads2k

### CODE REVIEW

**Comment by deads2k** [comment]
@fabianofranz

**Comment by fabianofranz** [comment]
@k8s-bot ok to test

**Comment by abdasgupta** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by abdasgupta** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by abdasgupta** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by abdasgupta** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by abdasgupta** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by erictune** [comment]
@kubernetes/sig-cli-pr-reviews

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go [hunk 1]
```diff
@@ -128,6 +128,12 @@ func (m *DefaultRESTMapper) Add(kind schema.GroupVersionKind, scope RESTScope) {
 	m.kindToScope[kind] = scope
 }
 
+// Specifically written for test cases to perform tests on resources.
+func (m *DefaultRESTMapper) AddCustomResource(singular, plural schema.GroupVersionResource) {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go]: I wouldn't expect exported functions for test purposes like this

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 2]
```diff
@@ -104,37 +109,64 @@ func (e shortcutExpander) getShortcutMappings() ([]kubectl.ResourceShortcuts, er
 	return res, nil
 }
 
+// resourceFinder will find the resources through discovery client as per resource plura name and singular name.
+func (e shortcutExpander) resourceFinder(resource schema.GroupVersionResource) (schema.GroupVersionResource, bool) {
+	apiResList, err := e.discoveryClient.ServerResources()
+	if err == nil {
+		for _, apiResources := range apiResList {
+			for _, apiRes := range apiResources.APIResources {
+				if apiRes.Name == resource.Resource {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: if resource.Group and/or resource.Version are non-empty, make sure those match apiRes as well

File: staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go (hunk 1 above)
Comment (abdasgupta) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go]: Any suggestion otherwise??

Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go]: create a discovery.NewRESTMapper and feed in the groups/versions/resources you want to test with as if they had been retrieved via a call to the discovery doc

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 2 above)
Comment (abdasgupta) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Hi liggitt,
They are already checked in "staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go". Do we need to check them again here before that?

Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: yes, otherwise, this could return a resource from the wrong group.

Comment (abdasgupta) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: @liggitt 
apiRes doesn't have Group or Version to match with -vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

It looks, only way to Group/Version match is to match with mapper which is actually happening in restmapper.go. Updated the PR.

Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: the group/version is in `apiResources.GroupVersion` in the outer loop

Comment (abdasgupta) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Done

**Pull request closed without merging**